### DISCUSSION TITLE
Kratos Constitutive laws

### DISCUSSION THREAD

**Discussion by AriasRH** [KratosMultiphysics/Kratos#12210] [Q&A]
Hola buenos días, Soy Heriberto de Aula CIMNE Morelia, Mex.
Tengo dudas respecto al uso de las leyes constitutivas. 

El problema es un análisis estructural 2D. Suelo modificar propiedades directamente en el JSON pero cuando intento implementar alguna de las leyes constitutivas que aparecen en github (https://github.com/KratosMultiphysics/Kratos/blob/master/applications/ConstitutiveLawsApplication/constitutive_laws_application.cpp)

me muestra el siguiente error:
kratos/utilities/read_materials_utility.cpp:205:ReadMaterialsUtility::AssignConstitutiveLawToProperty
kratos/utilities/read_materials_utility.cpp:212:ReadMaterialsUtility::AssignConstitutiveLawToProperty
 kratos/utilities/read_materials_utility.cpp:185:ReadMaterialsUtility::AssignMaterialToProperty
kratos/utilities/read_materials_utility.cpp:474:ReadMaterialsUtility::AssignPropertyBlock
 kratos/utilities/read_materials_utility.cpp:148:ReadMaterialsUtility::GetPropertyBlock
 kratos/utilities/read_materials_utility.cpp:82:ReadMaterialsUtility::ReadMaterialsUtility

No se si hay algún problema de compatibilidad entre la versión de Kratos que yo utilizo (9.2.0) y las leyes constitutivas de github (Parece que solo funcionan las que aparecen por default en GID 16.0)

**Comment by AlejandroCornejo** [comment]
Hola!

¿Qué ley en concreto quieres utilizar? Creo que el nombre no está bien puesto en tu json. Danos más info y tratamos de ayudarte. Si puedes usar el GiD 16.1.8d sería más fácil.

Alejandro

**Reply by AriasRH** [reply]
Hola!
Estoy probando algunas de estas. Copio el nombre de github. Solo estaba probando diferentes leyes para ver el comportamiento del modelo.
KRATOS_REGISTER_CONSTITUTIVE_LAW("SmallStrainIsotropicPlasticityPlaneStrainVonMisesVonMises", mSmallStrainIsotropicPlasticityPlaneStrainVonMisesVonMises);
    KRATOS_REGISTER_CONSTITUTIVE_LAW("SmallStrainIsotropicPlasticityPlaneStrainVonMisesModifiedMohrCoulomb", mSmallStrainIsotropicPlasticityPlaneStrainVonMisesModifiedMohrCoulomb);
    KRATOS_REGISTER_CONSTITUTIVE_LAW("SmallStrainIsotropicPlasticityPlaneStrainVonMisesDruckerPrager", mSmallStrainIsotropicPlasticityPlaneStrainVonMisesDruckerPrager);
    KRATOS_REGISTER_CONSTITUTIVE_LAW("SmallStrainIsotropicPlasticityPlaneStrainVonMisesTresca", mSmallStrainIsotropicPlasticityPlaneStrainVonMisesTresca);
    KRATOS_REGISTER_CONSTITUTIVE_LAW("SmallStrainIsotropicPlasticityPlaneStrainModifiedMohrCoulombVonMises", mSmallStrainIsotropicPlasticityPlaneStrainModifiedMohrCoulombVonMises);
    KRATOS_REGISTER_CONSTITUTIVE_LAW("SmallStrainIsotropicPlasticityPlaneStrainModifiedMohrCoulombModifiedMohrCoulomb", mSmallStrainIsotropicPlasticityPlaneStrainModifiedMohrCoulombModifiedMohrCoulomb);
    KRATOS_REGISTER_CONSTITUTIVE_LAW("SmallStrainIsotropicPlasticityPlaneStrainModifiedMohrCoulombDruckerPrager", mSmallStrainIsotropicPlasticityPlaneStrainModifiedMohrCoulombDruckerPrager);
    KRATOS_REGISTER_CONSTITUTIVE_LAW("SmallStrainIsotropicPlasticityPlaneStrainModifiedMohrCoulombTresca"

[ProjectParameters.json](https://github.com/KratosMultiphysics/Kratos/files/14669224/ProjectParameters.json)
[StructuralMaterials.json](https://github.com/KratosMultiphysics/Kratos/files/14669225/StructuralMaterials.json)

Lo raro es que cuando realizo pruebas de un modelo 3D, si funcionan todas las leyes constitutivas.
También me cuesta un poco de trabajo identificar las variables que involucran cada una de las leyes. 
Quiero entender todos los parámetros y opciones de los archivos JSON

Saludos y gracias por responder tan rápido.

**Comment by AriasRH** [comment]
Hola
Estoy probando algunas de estas. Copio el nombre de github. Solo estaba
probando diferentes leyes para ver el comportamiento del modelo.
Lo raro es que cuando realizo pruebas de un modelo 3D, si funcionan todas
las leyes constitutivas.
También me cuesta un poco de trabajo identificar las variables que
involucran cada una de las leyes.
Quiero entender todos los parámetros y opciones de los archivos JSON
[image: image.png]
Saludos y gracias por responder tan rápido.

**Comment by PetitPipi** [comment]
Hi,
I also ran into this kind of problem when I was trying out the FEM-MPM coupled example in CoSimulationApplication.
It seems like the HyperElasticPlaneStrain2DLaw model is not in the Structural Mechanics Application (FEM)? 
So, first I wanna make sure that the name of the model was not wrong.  If the model is truly missing, what can I do to include this model?
The error message is as follow:

> RuntimeError: Error: Kratos components missing "HyperElasticPlaneStrain2DLaw"
> 
> in kratos/utilities/read_materials_utility.cpp:209: ReadMaterialsUtility::AssignConstitutiveLawToProperty
>    kratos/utilities/read_materials_utility.cpp:216: ReadMaterialsUtility::AssignConstitutiveLawToProperty
>    kratos/utilities/read_materials_utility.cpp:189: ReadMaterialsUtility::AssignMaterialToProperty
>    kratos/utilities/read_materials_utility.cpp:493: ReadMaterialsUtility::AssignPropertyBlock
>    kratos/utilities/read_materials_utility.cpp:149: ReadMaterialsUtility::GetPropertyBlock
>    kratos/utilities/read_materials_utility.cpp:83: ReadMaterialsUtility::ReadMaterialsUtility

**Reply by rubenzorrilla** [reply]
`HyperElasticPlaneStrain2DLaw` is in the `ConstitutiveLawsApplication`.

**Reply by PetitPipi** [reply]
Thanks, It worked!
